File: backend/app/schemas/payment.py
from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, Field

from app.schemas.money import MoneyAmount, OptionalMoneyAmount


class PaymentStatus(StrEnum):
    pending = "pending"
    processing = "processing"
    completed = "completed"
    failed = "failed"
    cancelled = "cancelled"


class PaymentMethod(StrEnum):
    ach = "ach"
    wire = "wire"
    check = "check"
    virtual_card = "virtual_card"
    # UK domestic rails. A GBP→GB domestic payment used to fall through to
    # `international_wire` (SWIFT + a 2.5% fee anchor) because the corridor
    # selector had no domestic-GB branch and none of these rails existed;
    # `payment_corridor.pick_corridor` now routes a same-currency GBP/GB
    # payment onto Faster Payments, with BACS / CHAPS available as explicit
    # `requested_method` overrides (issue #328). All three are domestic bank
    # rails — 1099-reportable, not card, no FX leg.
    bacs = "bacs"
    faster_payments = "faster_payments"
    chaps = "chaps"


class PaymentRunStatus(StrEnum):
    """Every status `PaymentRun.status` can actually hold.

    `executing`, `partial` and `cancelled` were missing even though the code
    has always written them: `/runs/{id}/execute` claims a run by flipping it
    to `executing`, `_dispatch_run_payments` rolls up to `partial` when some
    payments succeeded and some failed, and `/runs/{id}/cancel` writes
    `cancelled`. An enum that can't name three of the eight real values is a
    filter/validation surface that silently disagrees with the table.
    """

    draft = "draft"
    executing = "executing"
    submitted = "submitted"
    processing = "processing"
    partial = "partial"
    completed = "completed"
    failed = "failed"
    cancelled = "cancelled"


PAYMENT_STATUSES = [s.value for s in PaymentStatus]
PAYMENT_RUN_STATUSES = [s.value for s in PaymentRunStatus]


class PaymentCreate(BaseModel):
    invoice_id: str
    # Optional, and only ever a cross-check: the server binds the payment to
    # the invoice amount net of applied credit memos and 422s a value that
    # disagrees (see `api/payments.create_payment`). Required-ness here was a
    # trap once netting landed — a caller that knows the invoice amount but not
    # its credits had no figure it could legally send.
    # Digits match `payments.amount` Numeric(15, 2).
    amount: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    method: PaymentMethod | None = None
    reference: str | None = Field(default=None, max_length=255)
    # No `payment_run_id`. It used to be accepted here and written straight to
    # the FK with no validation at all — the run was never checked to exist, to
    # be `draft`, or to belong to the caller's entity, and `run.total_amount` /
    # `requires_cfo_approval` were not recomputed. That let a caller inject
    # payments into any run: N legs each individually under
    # `payments.cfo_approval_above` inflate a run whose CFO flag was frozen at
    # creation, and `/execute` then dispatches the lot with no sign-off. It also
    # let a payment be attached to an already-terminal run, where nothing ever
    # dispatches it — leaving the row `pending` forever, occupying the invoice's
    # `uq_payments_one_live_per_invoice` slot with `/void` the only exit.
    #
    # A payment that belongs to a run is created BY the run
    # (`services/payment_runs.create_payment_run_for_invoices`), which stamps
    # the FK itself. This endpoint books a STANDALONE payment and gates it with
    # its own per-payment CFO check; there is no legitimate caller. No
    # first-party client ever sent the field, and Pydantic ignores an unknown
    # key, so a stray one is simply not honoured rather than 422-ing.


class PaymentResponse(BaseModel):
    id: str
    correlation_id: str | None
    invoice_id: str
    payment_run_id: str | None
    amount: MoneyAmount
    method: str | None
    status: str
    reference: str | None
    created_at: str
    updated_at: str | None

    # Why this payment failed, and when it moved. `Payment.failure_reason` has
    # existed (and been populated on every failure path — compliance refusal,
    # card-issuance failure, adapter error, void, webhook failure) since the
    # model was written, but it never reached the read surface: a run could
    # report "2 failed" and the UI had no way to say why, so the operator's
    # only recourse was the server log. Readable by the payments roles
    # (admin / ap_manager / cfo), who can already read vendor bank details —
    # this is not a new exposure class.
    provider: str | None = None
    failure_reason: str | None = None
    submitted_at: str | None = None
    completed_at: str | None = None

    # What the PROCESSOR says it moved, beside `amount` which is what AP
    # AUTHORIZED (migration 0083). Without these on the read surface an
    # operator seeing a `completed` payment whose invoice is stuck at
    # `payment_scheduled` has no way to tell why — the shortfall would only be
    # visible by cross-referencing the audit log or a `fraud_flag` exception
    # that doesn't name the figures. Same reasoning as `failure_reason` above,
    # and the same audience.
    #
    # `None` here means no rail ever reported a figure, NOT zero — see
    # `payment_settlement.settlement_coverage`.
    settled_amount: OptionalMoneyAmount = None
    settled_currency: str | None = None

    # Joined fields from invoice
    vendor_name: str | None = None
    invoice_number: str | None = None

    # Card metadata: surfaced when method=virtual_card so the History row
    # can show "•••• 1234 · lithic" without an extra request.
    card_last_four: str | None = None
    card_provider: str | None = None
    card_id: str | None = None

    model_config = {"from_attributes": True}

    @classmethod
    def from_db(cls, p, invoice=None, card=None) -> "PaymentResponse":
        return cls(
            id=str(p.id),
            correlation_id=str(p.correlation_id) if p.correlation_id else None,
            invoice_id=str(p.invoice_id),
            payment_run_id=str(p.payment_run_id) if p.payment_run_id else None,
            amount=p.amount,
            method=p.method,
            status=p.status,
            reference=p.reference,
            created_at=p.created_at.isoformat() if p.created_at else "",
            updated_at=p.updated_at.isoformat() if p.updated_at else None,
            provider=p.provider,
            failure_reason=p.failure_reason,
            submitted_at=p.submitted_at.isoformat() if p.submitted_at else None,
            completed_at=p.completed_at.isoformat() if p.completed_at else None,
            settled_amount=p.settled_amount,
            settled_currency=p.settled_currency,
            vendor_name=invoice.vendor_name if invoice else None,
            invoice_number=invoice.invoice_number if invoice else None,
            card_last_four=card.last_four if card else None,
            card_provider=card.card_provider if card else None,
            card_id=str(card.id) if card else None,
        )


class PaymentListResponse(BaseModel):
    items: list[PaymentResponse]
    total: int
    page: int
    page_size: int


class PaymentRunResponse(BaseModel):
    id: str
    status: str
    total_amount: OptionalMoneyAmount = None
    initiated_by: str | None
    executed_at: str | None
    created_at: str
    payment_count: int = 0

    # Per-outcome tallies over the run's own payments. `partial` alone doesn't
    # tell an operator whether one payment failed or forty, and the counts used
    # to exist only in the transient response body of the /execute call that
    # produced them — reload the page and they were gone. Derived on read from
    # the child `Payment` rows (no stored running total, matching the budget
    # service's compute-on-read posture), so they can never drift from the
    # payments they summarise.
    payments_completed: int = 0
    payments_failed: int = 0
    payments_in_flight: int = 0
    payments_pending: int = 0

    # The CFO sign-off gate, on the LIST shape. These columns have always
    # existed on the row and `GET /runs/{id}` (a raw dict) has always returned
    # them — but the list endpoint declares this model, and FastAPI strips
    # whatever a response model doesn't declare. A client that reads the list
    # therefore saw `requires_cfo_approval` as absent for every run, always,
    # and could not tell an above-threshold run from any other: the mobile app
    # renders no "CFO approval required" marker, its pre-flight gate evaluates
    # false, and Execute goes out to a 403. The web app is unaffected only
    # because its run modal happens to read the detail endpoint instead.
    requires_cfo_approval: bool = False
    cfo_approved_at: str | None = None

    model_config = {"from_attributes": True}

    @classmethod
    def from_db(
        cls,
        pr,
        payment_count: int = 0,
        *,
        completed: int = 0,
        failed: int = 0,
        in_flight: int = 0,
        pending: int = 0,
    ) -> "PaymentRunResponse":
        return cls(
            id=str(pr.id),
            status=pr.status,
            total_amount=pr.total_amount,
            initiated_by=str(pr.initiated_by) if pr.initiated_by else None,
            executed_at=pr.executed_at.isoformat() if pr.executed_at else None,
            created_at=pr.created_at.isoformat() if pr.created_at else "",
            payment_count=payment_count,
            payments_completed=completed,
            payments_failed=failed,
            payments_in_flight=in_flight,
            payments_pending=pending,
            requires_cfo_approval=bool(getattr(pr, "requires_cfo_approval", False)),
            cfo_approved_at=(
                pr.cfo_approved_at.isoformat() if getattr(pr, "cfo_approved_at", None) else None
            ),
        )


class PaymentRunListResponse(BaseModel):
    items: list[PaymentRunResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/lib/types/payment.ts
import type { BadgeTone } from '$lib/components/ui/Badge.svelte';
import type { MessageKey } from '$lib/i18n/messages';
import type { MoneyAmount } from '$lib/utils/money';

// Mirrors the statuses the backend actually persists on `payments.status`.
// `pending_compliance` is the parking state the sanctions/KYC gate
// (`services/compliance.check_payment_compliance`) leaves a payment in — it
// must be listed here or the History badge renders blank (no label) and the
// status has no filter chip, which is how a held payment stayed invisible.
// Its two exits are `POST /api/payments/{id}/compliance/{release,dismiss}`.
export type PaymentStatus =
	| 'pending'
	| 'pending_compliance'
	| 'submitted'
	| 'processing'
	| 'completed'
	| 'failed'
	| 'cancelled'
	| 'voided';

export const PAYMENT_STATUSES: PaymentStatus[] = [
	'pending',
	'pending_compliance',
	'submitted',
	'processing',
	'completed',
	'failed',
	'cancelled',
	'voided'
];

/**
 * The i18n key carrying each status label — never the English string itself.
 *
 * Both surfaces that render a payment status (the `/payments` History badge
 * and its filter chips) are inside the i18n extraction slice, so a hardcoded
 * English map here put a translated chip row directly beside an untranslated
 * `Compliance Hold` badge. Keyed the same way
 * `notification.ts::EVENT_LABEL_KEYS` is; `Record<PaymentStatus, MessageKey>`
 * makes a new status a compile error rather than a blank badge, and
 * `paymentStatus.test.ts` proves every key exists in the catalogue.
 */
export const PAYMENT_STATUS_LABEL_KEYS: Record<PaymentStatus, MessageKey> = {
	pending: 'payments.status.pending',
	pending_compliance: 'payments.status.pendingCompliance',
	submitted: 'payments.status.submitted',
	processing: 'payments.status.processing',
	completed: 'payments.status.completed',
	failed: 'payments.status.failed',
	cancelled: 'payments.status.cancelled',
	voided: 'payments.status.voided'
};

/**
 * The message key for a status, or `null` for one this frontend doesn't know
 * (the caller renders the raw value — visible and searchable — rather than a
 * blank badge). `Payment.status` is typed against the union, but a status the
 * backend adds first still reaches the badge as a bare string.
 */
export function paymentStatusLabelKey(status: string): MessageKey | null {
	return PAYMENT_STATUS_LABEL_KEYS[status as PaymentStatus] ?? null;
}

/**
 * Badge tone per payment status.
 *
 * Lives here rather than on `/payments` because the same union is badged in
 * more than one place — the History table, the queue, and `RunDetailModal`'s
 * per-payment column — and the two used to disagree: the modal tinted
 * `pending` amber but had no rule at all for `submitted`, `cancelled`,
 * `voided` or `pending_compliance`, so half the union rendered untinted there
 * while `/payments` painted it. That is the drift a shared map exists to stop
 * (`frontend/CLAUDE.md` § Badge).
 *
 * `Record<PaymentStatus, …>` on purpose: the old per-status CSS rules were a
 * list you had to remember to extend, and `voided` — a real member of
 * `PAYMENT_STATUSES` — never got one, so a voided payment rendered an
 * untinted pill. A total record makes the compiler ask the question.
 *
 * `cancelled` and `voided` share `muted` (a grey tint) while a run's `draft`
 * is `neutral` (flat): "abandoned" and "reversed" are both inert money states,
 * whereas draft is money that has not been attempted yet.
 *
 * No `?? 'neutral'` at the call sites, unlike {@link RUN_STATUS_TONES} below:
 * this record is total over the union, and a status off the union lands on
 * `Badge`'s own `tone` default — which is `neutral` — rather than on a
 * fallback restating it.
 */
export const PAYMENT_STATUS_TONES: Record<PaymentStatus, BadgeTone> = {
	pending: 'warning',
	// Same tone as `pending` — both are waiting. What separates "a human must
	// clear this" from "waiting its turn" is the ring the History cell draws
	// around it (`/payments` `.compliance-ring`), a caller-owned wrapper
	// rather than a sixth tone — decisions.md §52.
	pending_compliance: 'warning',
	submitted: 'accent',
	processing: 'accent',
	completed: 'success',
	failed: 'danger',
	cancelled: 'muted',
	voided: 'muted'
};

/**
 * The statuses a payment RUN can report — `services/payment_runs`' three CLAIM
 * states (`draft` / `executing` / `cancelled`, passed through untouched) plus
 * the four `PaymentRunRollup.run_status` derives from the run's own payments
 * (`submitted` / `completed` / `partial` / `failed`). Mirrors
 * `backend/app/services/payment_runs.py`; decisions.md §41.
 *
 * `PaymentRun.status` stays a bare `string` on the wire — the value is derived
 * on read, so a future rollup rung reaches the UI before this union does. The
 * union exists to make the two maps below TOTAL over what the backend can
 * report today: `partial` and `executing` had no tone rule at all and rendered
 * untinted, and neither had a label.
 */
export type PaymentRunStatus =
	| 'draft'
	| 'executing'
	| 'submitted'
	| 'completed'
	| 'partial'
	| 'failed'
	| 'cancelled';

export const PAYMENT_RUN_STATUSES: PaymentRunStatus[] = [
	'draft',
	'executing',
	'submitted',
	'completed',
	'partial',
	'failed',
	'cancelled'
];

/**
 * Badge tone per payment-RUN status — `services/payment_runs`' three claim
 * states (`draft` / `executing` / `cancelled`) plus the four its rollup
 * derives (decisions.md §41).
 *
 * Total over {@link PaymentRunStatus}, so every status the backend can report
 * today is tinted. `PaymentRun.status` is still a bare string on the wire (the
 * backend derives it on read), so read it through {@link runStatusTone}, which
 * falls an unknown value back to the flat `neutral` chip rather than to
 * nothing. `partial` and `executing` had no rule at all on either surface and
 * rendered untinted — `partial` especially, which is the one run status
 * meaning "some of this money failed".
 *
 * `draft` is `neutral`, not amber: `RunDetailModal` tinted it `warning` while
 * `/payments` rendered it flat, for the same run, one click apart. Flat wins —
 * a draft run is money that has not been attempted yet, which is the absence
 * of a signal rather than a weak one.
 */
export const RUN_STATUS_TONES: Record<PaymentRunStatus, BadgeTone> = {
	draft: 'neutral',
	executing: 'accent',
	submitted: 'accent',
	completed: 'success',
	partial: 'warning',
	failed: 'danger',
	cancelled: 'muted'
};

export function runStatusTone(status: string): BadgeTone {
	return RUN_STATUS_TONES[status as PaymentRunStatus] ?? 'neutral';
}

/**
 * The i18n key carrying each run status' label — never the English string
 * itself.
 *
 * `PaymentRun.status` had no label map at all, so both surfaces that badge it
 * — the `/payments` Runs table and `RunDetailModal`'s header — printed the RAW
 * enum, one cell away from a per-payment badge that round 24 had keyed. A
 * German user read `executing` in Latin script inside an otherwise translated
 * dialog.
 *
 * Keyed off the same union as {@link RUN_STATUS_TONES}, so a status with a
 * tone but no label (a coloured pill showing a raw value) is a compile error;
 * `paymentStatus.test.ts` proves every key exists in the catalogue.
 */
export const RUN_STATUS_LABEL_KEYS: Record<PaymentRunStatus, MessageKey> = {
	draft: 'paymentRuns.status.draft',
	executing: 'paymentRuns.status.executing',
	submitted: 'paymentRuns.status.submitted',
	completed: 'paymentRuns.status.completed',
	partial: 'paymentRuns.status.partial',
	failed: 'paymentRuns.status.failed',
	cancelled: 'paymentRuns.status.cancelled'
};

/**
 * The message key for a run status, or `null` for one this frontend doesn't
 * know (the caller renders the raw value — visible and searchable — rather
 * than a blank badge). `PaymentRun.status` is a bare string on the wire: the
 * backend DERIVES it on read, so a rollup rung this build has never seen
 * arrives as ordinary text.
 */
export function runStatusLabelKey(status: string): MessageKey | null {
	return RUN_STATUS_LABEL_KEYS[status as PaymentRunStatus] ?? null;
}

// UK domestic bank rails (bacs / faster_payments / chaps). A same-currency
// GBP payment to a GB vendor stays inside the UK banking system (sort code +
// account number — no IBAN, no SWIFT, no FX); the backend corridor selector
// (`payment_corridor.pick_corridor`) auto-selects `faster_payments` and honours
// an explicit `bacs` / `chaps` override. See issue #328.
export type PaymentMethod =
	| 'ach'
	| 'wire'
	| 'check'
	| 'virtual_card'
	| 'bacs'
	| 'faster_payments'
	| 'chaps';

export const PAYMENT_METHODS: PaymentMethod[] = [
	'ach',
	'wire',
	'check',
	'virtual_card',
	'bacs',
	'faster_payments',
	'chaps'
];

/**
 * The i18n key carrying each rail's label — never the English string itself.
 * The rail is a column on the (extracted) `/payments` History table and in
 * `RunDetailModal`, one cell from the now-translated status badge.
 *
 * Several rails are proper NAMES rather than words — `ACH`, `BACS`, `Faster
 * Payments`, `CHAPS` are scheme names and their catalogue entries are the
 * same string in all six locales. That is deliberate: routing them through a
 * key still costs nothing, and it means the day a locale does want to
 * transliterate one, there is a place to put it.
 */
export const PAYMENT_METHOD_LABEL_KEYS: Record<PaymentMethod, MessageKey> = {
	ach: 'payments.method.ach',
	wire: 'payments.method.wire',
	check: 'payments.method.check',
	virtual_card: 'payments.method.virtualCard',
	bacs: 'payments.method.bacs',
	faster_payments: 'payments.method.fasterPayments',
	chaps: 'payments.method.chaps'
};

/**
 * The message key for a rail, or `null` for one this frontend doesn't know
 * (the caller renders the raw value — visible and searchable — rather than a
 * blank cell). `Payment.method` is nullable on the wire and the backend can
 * add a rail before this union does.
 */
export function paymentMethodLabelKey(method: string): MessageKey | null {
	return PAYMENT_METHOD_LABEL_KEYS[method as PaymentMethod] ?? null;
}

export interface Payment {
	id: string;
	correlation_id: string | null;
	invoice_id: string;
	payment_run_id: string | null;
	/** `schemas/payment.py::PaymentResponse.amount` is `MoneyAmount` — a JSON number on the wire. */
	amount: MoneyAmount;
	method: PaymentMethod | null;
	status: PaymentStatus;
	reference: string | null;
	created_at: string;
	updated_at: string | null;
	/**
	 * What the PROCESSOR says it settled, beside `amount` — what AP AUTHORIZED
	 * (`schemas/payment.py::PaymentResponse`, migration 0083). Both have been on
	 * the read surface since the settlement-verification work; nothing rendered
	 * them, which is how an invoice held at `payment_scheduled` for an
	 * under-settlement was invisible in the app.
	 *
	 * `null` means no rail ever reported a figure — NOT zero. The backend fails
	 * OPEN on that case (`payment_settlement.settlement_coverage`), so absence
	 * is never evidence of a shortfall. Whether a reported figure actually
	 * covers the invoice is the SERVER's call: never subtract or compare these
	 * two client-side (frontend/CLAUDE.md § Money formatting).
	 */
	settled_amount?: MoneyAmount;
	settled_currency?: string | null;
	/**
	 * What `amount` — the AUTHORIZED figure — is denominated in
	 * (`schemas/payment.py::PaymentResponse.currency`), off the invoice row the
	 * response already joins. `payments` has no currency column; a payment
	 * settles in its invoice's currency.
	 *
	 * Without it, six `/payments` call sites rendered the authorized amount
	 * under the ORG's default code. The sharpest was the Accept-settlement
	 * dialog, whose whole job is to put "Authorized" beside "Settled": the
	 * settled half had `settled_currency` and the authorized half had nothing,
	 * so a EUR payment showed a fabricated `$1,200.00` above a real `€1,150.00`
	 * on the screen built to catch a `currency_mismatch`.
	 *
	 * `null` means the server could not establish it — render the bare figure,
	 * never a substituted default (`docs/decisions.md` §79/§82).
	 */
	currency?: string | null;
	vendor_name: string | null;
	invoice_number: string | null;
	card_last_four: string | null;
	card_provider: string | null;
	card_id: string | null;
}

export interface PaymentRun {
	id: string;
	status: string;
	total_amount: MoneyAmount;
	/**
	 * What `total_amount` is denominated in
	 * (`schemas/payment.py::PaymentRunResponse.currency`). `payment_runs` has no
	 * currency column either — the total is one bare `Numeric`, kept meaningful
	 * by `create_payment_run_for_invoices` refusing a run whose invoices span
	 * more than one currency.
	 *
	 * `null` when the server could not PROVE one: a run with no payments, or a
	 * legacy run predating that guard whose legs disagree — in which case the
	 * total is denominated in nothing real and a code would be worse than none.
	 * Render the bare figure (`docs/decisions.md` §79/§82).
	 */
	currency?: string | null;
	initiated_by: string | null;
	executed_at: string | null;
	created_at: string;
	payment_count: number;
}
